pkg/config/expression.go

```diff
@@ -64,47 +67,216 @@ func SimpleVarToReference(s string) (Reference, error) {
 
 	case 3:
 		return Reference{
-			Group:  components[0],
 			Module: components[1],
-			Name:   components[2]}, nil
+			Name:   components[2]}, components[0], nil
 	default:
-		return Reference{}, fmt.Errorf(
+		return Reference{}, "", fmt.Errorf(
 			"expected either 2 or 3 components, got %d in %#v", len(components), s)
 	}
 }
 
-// VariableTranslator is an interface that provides function
-// to translate "simple" variable (`$(...)`) into HCL format
-type VariableTranslator interface {
-	TranslateSimpleToHcl(s string) (string, error)
+// SimpleVarToHclExpression takes a string `$(...)` and transforms it to `HclExpression`
+// Returns `HclExpression` and name of explicit group if one was set.
+func SimpleVarToHclExpression(s string) (HclExpression, string, error) {
+	ref, gr, err := SimpleVarToReference(s)
+	if err != nil {
+		return HclExpression{}, "", err
+	}
+	var ex HclExpression
+	if ref.GlobalVar {
+		ex, err = ParseExpression(fmt.Sprintf("var.%s", ref.Name))
+	} else {
+		ex, err = ParseExpression(fmt.Sprintf("module.%s.%s", ref.Module, ref.Name))
+	}
+	if err != nil {
+		return HclExpression{}, "", err
+	}
+	return ex, gr, nil
+}
+
+// TraversalToReference takes HCL traversal and returns `Reference`
+func TraversalToReference(t hcl.Traversal) (Reference, error) {
+	if t.IsRelative() {
+		return Reference{}, fmt.Errorf("got relative traversal")
+	}
+	getAttrName := func(i int) (string, error) {
+		if i >= len(t) {
+			return "", fmt.Errorf("traversal does not have enough components")
+		}
+		if a, ok := t[i].(hcl.TraverseAttr); ok {
+			return a.Name, nil
+		}
+		return "", fmt.Errorf("got unexpected traversal component: %#v", t[i])
+	}
+	switch root := t.RootName(); root {
+	case "var":
+		n, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of global var reference to be a variable name, got %w", err)
+		}
+		return Reference{GlobalVar: true, Name: n}, nil
+	case "module":
+		m, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of module var reference to be a module name, got %w", err)
+		}
+		n, err := getAttrName(2)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected third component of module var reference to be a variable name, got %w", err)
+		}
+		return Reference{Module: m, Name: n}, nil
+	default:
+		return Reference{}, fmt.Errorf("unexpected first component of reference: %#v", root)
+	}
+}
+
+// IsRawHclLiteral checks if passed value of type cty.String
+// and its content starts with "((" and ends with "))".
+// Returns trimmed string and result of test.
+func IsRawHclLiteral(v cty.Value) (string, bool) {
+	if v.Type() != cty.String {
+		return "", false
+	}
+	s := v.AsString()
+	if len(s) < 4 || s[:2] != "((" || s[len(s)-2:] != "))" {
+		return "", false
+	}
+	return s[2 : len(s)-2], true
+}
+
+// HclExpression is a representation of HCL literals in Blueprint
+type HclExpression struct {
```

Comment:
I realize that these are private fields, but I think it is worth adding a comment that `s` is a string in HCL syntax. e.g. a static value, a "var.x" reference or a "module.x.y" reference.

If these fields are used directly anywhere in this package, I believe they should have longer, meaningful variable names. If they are uniformly accessed through getters, I am less concerned.

Reply:
Those fields meant to be "private"

Reply:
OK: I am comfortable with the variables names, but do please add comment. I realize we are getting toward a simpler world, but the multiple string representations of things can make it complicated to know "which is which and where".